src/main/java/dev/openfga/sdk/api/client/HttpRequestAttempt.java

```diff
@@ -0,0 +1,106 @@
+                .sendAsync(request, HttpResponse.BodyHandlers.ofString())
+                .thenCompose(response -> {
+                    try {
+                        checkStatus(name, response, previousError);
+                    } catch (FgaApiRateLimitExceededError | FgaApiInternalError retryableError) {
+                        if (retryNumber < configuration.getMaxRetries()) {
+                            HttpClient delayingClient = getDelayedHttpClient();
+                            return attemptHttpRequest(delayingClient, retryNumber + 1, retryableError);
+                        }
+                        return CompletableFuture.failedFuture(retryableError);
+                    } catch (ApiException e) {
+                        return CompletableFuture.failedFuture(e);
+                    }
+
+                    return deserializeResponse(response);
+                });
+    }
+
+    private CompletableFuture<T> deserializeResponse(HttpResponse<String> response) {
+        if (clazz == Void.class && isNullOrWhitespace(response.body())) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        try {
+            T deserialized = apiClient.getObjectMapper().readValue(response.body(), clazz);
+            return CompletableFuture.completedFuture(deserialized);
+        } catch (IOException e) {
+            // Malformed response.
+            return CompletableFuture.failedFuture(new ApiException(e));
+        }
+    }
+
+    private HttpClient getDelayedHttpClient() {
+        Duration retryDelay = configuration.getMinimumRetryDelay();
+        return apiClient
+                .getHttpClientBuilder()
+                .executor(CompletableFuture.delayedExecutor(retryDelay.toNanos(), TimeUnit.NANOSECONDS))
+                .build();
+    }
+
+    private static void checkStatus(String name, HttpResponse<String> response, Throwable previousError)
+            throws ApiException {
+        int status = response.statusCode();
+        String body = response.body();
+
+        switch (status) {
+            case HttpURLConnection.HTTP_BAD_REQUEST:
+            case 422: // HTTP 422 Unprocessable Entity
+                throw new FgaApiValidationError(name, previousError, status, response.headers(), body);
+
+            case HttpURLConnection.HTTP_UNAUTHORIZED:
+            case HttpURLConnection.HTTP_FORBIDDEN:
+                throw new FgaApiAuthenticationError(name, previousError, status, response.headers(), body);
+
+            case HttpURLConnection.HTTP_NOT_FOUND:
+                throw new FgaApiNotFoundError(name, previousError, status, response.headers(), body);
+
+            case 429: // HTTP 429 Too Many Requests
+                throw new FgaApiRateLimitExceededError(name, previousError, status, response.headers(), body);
+
+            case HttpURLConnection.HTTP_INTERNAL_ERROR:
+                throw new FgaApiInternalError(name, previousError, status, response.headers(), body);
+        }
```

Comment:
do we need some form of default if it does not fit any?

Reply:
Good point

Let me see if there's a way from mustache to know the expected return code (200, 201, 204) and assert on that.

If not I can just add a final check to require it's a 2xx return code

Reply:
Couldn't find a way to get "a single response status code" out of OpenAPI generator or mustache. I think it's not possible since OpenAPI doesn't really differentiate "ok" responses from "error" responses

I'm adding a check to throw if the exception is not HTTP 2XX

[b9863d5](https://github.com/openfga/sdk-generator/pull/217/commits/b9863d5791b7855262fee8dcfbc13f9d8097ddae)